File: 프로그래머스/2/160585. 혼자서 하는 틱택토/혼자서 하는 틱택토.cpp
#include <string>
#include <vector>

using namespace std;

static bool win(const vector<string>& b, char c) {
    // rows
    for (int i = 0; i < 3; i++) {
        if (b[i][0] == c && b[i][1] == c && b[i][2] == c) return true;
    }
    // cols
    for (int j = 0; j < 3; j++) {
        if (b[0][j] == c && b[1][j] == c && b[2][j] == c) return true;
    }
    // diagonals
    if (b[0][0] == c && b[1][1] == c && b[2][2] == c) return true;
    if (b[0][2] == c && b[1][1] == c && b[2][0] == c) return true;
    return false;
}

int solution(vector<string> board) {
    int oCnt = 0, xCnt = 0;
    for (auto &row : board) {
        for (char ch : row) {
            if (ch == 'O') oCnt++;
            else if (ch == 'X') xCnt++;
        }
    }

    // 1) count rule
    if (!(oCnt == xCnt || oCnt == xCnt + 1)) return 0;

    bool oWin = win(board, 'O');
    bool xWin = win(board, 'X');

    // 2) both win impossible
    if (oWin && xWin) return 0;

    // 3) winner vs count consistency
    if (oWin) {
        if (oCnt != xCnt + 1) return 0;
    }
    if (xWin) {
        if (oCnt != xCnt) return 0;
    }

    return 1;
}

File: 프로그래머스/2/160585. 혼자서 하는 틱택토/혼자서 하는 틱택토.java
class Solution {
    public int solution(String[] board) {        
        int o = 0;
        int x = 0;
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length(); j++) {
                if (board[i].charAt(j) == 'O') o++;
                else if (board[i].charAt(j) == 'X') x++;
            }
        }
        
        if (o != x && o != x + 1) return 0;
        boolean oWin = line(board, 'O');
        boolean xWin = line(board, 'X');
        
        if (oWin && xWin) return 0;
        if (oWin && o != x + 1) return 0;
        if (xWin && o != x) return 0;
        return 1;
    }
    
    private boolean line(String[] board, char c) {
        for (int i = 0; i < 3; i++) {
            int cnt = 0;
            for (int j = 0; j < 3; j++) {
                if (board[i].charAt(j) == c) cnt++;
            }
            if (cnt == 3) return true;
        }
        for (int i = 0; i < 3; i++) {
            int cnt = 0;
            for (int j = 0; j < 3; j++) {
                if (board[j].charAt(i) == c) cnt++;
            }
            if (cnt == 3) return true;
        }
        int cnt = 0;
        for (int i = 0; i < 3; i++) {
            if (board[i].charAt(i) == c) cnt++;
        }
        if (cnt == 3) return true;
        cnt = 0;
        for (int i = 0; i < 3; i++) {
            if (board[i].charAt(2 - i) == c) cnt++;
        }
        if (cnt == 3) return true;
        return false;
    }
}